lib/membrane_mp4/demuxer/cmaf.ex

```diff
@@ -52,407 +52,244 @@ defmodule Membrane.MP4.Demuxer.CMAF do
   @impl true
   def handle_init(_ctx, _options) do
     state = %{
-      unprocessed_boxes: [],
-      unprocessed_binary: <<>>,
-      samples_info: nil,
-      track_to_pad_map: nil,
+      engine: __MODULE__.Engine.new(),
       all_pads_connected?: false,
-      buffered_actions: [],
-      fsm_state: :reading_cmaf_header,
-      track_notifications_sent?: false,
-      last_timescales: %{},
-      how_many_segment_bytes_read: 0,
-      tracks_info: nil,
-      tracks_notification_sent?: false
+      track_to_pad_map: nil,
+      new_tracks_sent?: false,
+      stream_format_sent?: false
     }
 
     {[], state}
   end
 
+  @impl true
+  def handle_pad_added(_pad, ctx, state) do
+    cond do
+      state.all_pads_connected? ->
+        raise "All pads have corresponding track already connected"
+
+      ctx.playback == :playing and not state.new_tracks_sent? ->
+        raise """
+        Pads can be linked either before #{inspect(__MODULE__)} enters :playing playback or after it \
+        sends {:new_tracks, ...} notification
+        """
+
+      true ->
+        :ok
+    end
+
+    state = %{state | all_pads_connected?: all_pads_connected?(ctx, state)}
```

Comment:
How about you wrapped that `all_pads_connected?` into some helper function that would:
* set the `all_pads_connected?` field in the state
* sent the `resume_auto_demand` action
* called `match_tracks_with_pads` function
* sent stream formats

?

Reply:
Moving `match_tracks_with_pads` to the point where all pads are connected would require changing the logic of `match_tracks_with_pads` and it would be even more complicated 😿